File: src/Cpp/Prime.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <cmath>
#include "Prime.h"

typedef long long llong;
using namespace std;

int lLength(llong l){
    llong  sizeTable[12] = {9, 99, 999, 9999, 99999, 999999, 9999999, 99999999, 999999999, 9999999999, 9999999999};
    for (int i = 0;  ; i++) if (l<= sizeTable[i]) return i+1;
}

string Prime::dfString(llong l) {
    string s = to_string(l);
    if (l >= 10000) s = s.substr(0, s.length() - 4) + "万";
    if (l % (10000*10000) == 0) s = s.substr(0, s.length() - 6) + "亿";
    else if (l >= 10000*10000) s = s.substr(0, s.length() - 6) + "亿" + s.substr(s.length() - 5);
    return s;
}

llong Prime::get(int index) {
    return index > _maxInd - 1 ? _prime[index - _offSet] : _prime[index];
}

llong Prime::size() {
    return  _maxInd;
}

void Prime::add(llong p) {
    _prime[_maxInd - _offSet] = p;
    _maxInd++;
}

void Prime::generateResults (llong inter, llong endNo){
    outputSequence(_prevNo,endNo);
    outputInterval(inter);
    _prevNo = endNo;
    freeUp();
}

void Prime::outputSequence(llong beginNo, llong endNo){
    string s;
    for (int i = lLength(beginNo) - 1; i <= lLength(endNo) - 1 ; i++) {
        for (int j = 1; j < 10 ; j++) {
            llong seq =  llong((j * pow(10, i)) + 0.5);
            if (seq < beginNo) continue;
            if (seq >= endNo) return;
            llong l = _prime[_maxInd - _offSet - 1 - (endNo - seq)];
            s = dfString(seq) + "|" + to_string(l) ;
            seqList.push_back(s);
            if (isDebug) printf("==>[No:] %s \n",s.c_str());
        }
    }
}

void Prime:: outputInterval(llong inter) {
    stringstream ss;
    if (inter % llong(pow(10, lLength(inter) - 1) + 0.5) == 0) {
        ss  << dfString(inter) << "|" << _maxInd << "|" << _prime[_maxInd - _offSet - 1] ;
        interList.push_back(ss.str());
        if (isDebug) cout << "[In:]" << ss.str() << endl;
    }
}

void Prime:: freeUp() {
    if (_maxInd > _maxKeep) _offSet = _maxInd-_maxKeep;
}

void Prime:: printTable(){
    cout << "## 素数区间表" << endl;
    cout <<"区间|个数|最大值"<< endl;
    cout <<"---|---|---"<< endl;
    for ( const string& s:interList) cout << s << endl;

    cout <<"## 素数序列表"<< endl;
    cout <<"序号|数值"<< endl;
    cout <<"---|---"<< endl;
    for ( const string& s:seqList) cout << s << endl;
}

Prime::Prime(llong limit) {
    _maxKeep = sqrt(limit) / log(sqrt(limit)) * 1.3 ;
    _prime = new llong[_maxKeep + 200000];
    cout << "内存使用：" << _maxKeep << endl;
}

File: original/Python/Prime.py
import math


class Prime:
    is_debug = True
    prime_list = []
    max_ind = 0
    max_keep = 0
    prev_no = 0
    seq_list = []
    inter_list = []

    def __init__(self, limit):
        self.max_keep = round(math.sqrt(limit) / math.log(math.sqrt(limit), math.e) * 1.5)

    def get(self, index):
        return self.prime_list[index] if (index < self.max_ind - 1) else self.prime_list[self.max_ind - 1]

    def getlast(self):
        return self.prime_list[len(self.prime_list) - 1]

    def size(self):
        return self.max_ind

    def add(self, p):
        self.prime_list.append(p)
        self.max_ind = self.max_ind + 1

    def generate_results(self, inter, end_no):
        self.output_sequence(self.prev_no, end_no)
        self.output_interval(inter)
        self.prev_no = end_no
        self.free_up()

    def output_interval(self, inter):
        if inter % pow(10, len(str(inter)) - 1) == 0:
            s = self.df_string(inter) + "|" + str(self.max_ind) + "|" + str(self.prime_list[len(self.prime_list) - 1])
            self.inter_list.append(s)
            if self.is_debug: print("[In:]" + s)

    def output_sequence(self, begin_no, end_no):
        for i in range(len(str(begin_no)) - 1, len(str(end_no))):
            for j in range(1, 10):
                seq = j * pow(10, i)
                if seq < begin_no: continue
                if seq >= end_no: return
                l = self.prime_list[len(self.prime_list) - 1 - (end_no - seq)]
                s = self.df_string(seq) + "|" + str(l)
                self.seq_list.append(s)
                if self.is_debug: print("==>[No:]" + s)

    def free_up(self):
        if self.max_ind > self.max_keep:
            del self.prime_list[self.max_keep:len(self.prime_list)]

    def df_string(self, l):
        s = str(l)
        if l % 1_0000_0000_0000 == 0:
            s = s[:-12] + "万亿"
        elif l % 1_0000_0000 == 0:
            s = s[:-8] + "亿"
        elif l % 1_0000 == 0:
            s = s[:-4] + "万"
        return s

    def print_table(self):
        print("## 素数区间表")
        print("区间|个数|最大值")
        print("---|---|---")
        for s in self.inter_list: print(s)
        print("## 素数序列表")
        print("序号|数值")
        print("---|---")
        for s in self.seq_list: print(s)
